### PR TITLE
Transitions -> Hooks

### PR DESCRIPTION
**Pull request by pavelradev812**
Transitions are now Hooks where, they run in the following sequence:
onExit(current state)->onEnter(next state)->animations

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -41,24 +41,29 @@ export default class Machine extends erx.Bus<State> {
...
     if (this.transitioning) {
       return false;
     }
     const finish = (next) => {
       this.state = next;
       this.push(next);
-      const eFn = this.entries[next];
-      if (eFn != null) {
-        next = eFn(this, data);
+      if (nextEntries != null && nextEntries.before) {
```
Comment (HDykins) [inline src/index.js]: I think the before function should be optional, and if so, animations should run whether nextEntries.before exists or not.

Comment (HDykins) [reply src/index.js]: Also shouldn't it be onEnter and onExit rather than before and after?

File: src/index.js [hunk 2]
```diff
@@ -41,24 +41,29 @@ export default class Machine extends erx.Bus<State> {
   }
 
   goTo(next: State, data?: any): boolean {
+    const currEntries = this.transitions[this.state] || null;
```
Comment (HDykins) [inline src/index.js]: We should remove transitions object now that we have hooks per entry? The entries should come from the entries table (although i think we committed to renaming entries to "states")

File: src/index.js [hunk 3]
```diff
@@ -41,24 +41,29 @@ export default class Machine extends erx.Bus<State> {
...
           this.goTo(next, data);
         }
       }
     };
-    const tFn = match(this.transitions, this.state, next);
-    if (tFn != null) {
+    if(currEntries != null && currEntries.after) {
       this.transitioning = true;
-      p(tFn(this, data)).then((change) => {
+      p(currEntries.after(this, data)).then(function (change) {
```
Comment (HDykins) [inline src/index.js]: This means that every onExit function must return truthy for the onEnter + animations of the next state to run. That was previously the behaviour of a transition object before, but do we want to keep that for onExit functions? Honestly don't know, what do you think?

**Review by HDykins — COMMENTED** [review]
Sorry it took a while, been a little busy today

File: src/index.js (hunk 1 above)
Comment (pavelradev812) [reply src/index.js]: You're right, my initial thoughts were that you'd always have animation/onEnter/onExit in an entry object, but it makes more sense to not have a property, than to have one with an empty value. Annd yes, should fix the naming

File: src/index.js (hunk 2 above)
Comment (pavelradev812) [reply src/index.js]: Yep the naming should change from transitions -> entries

File: src/index.js (hunk 3 above)
Comment (pavelradev812) [reply src/index.js]: Like the animations I had this thinking where, you'd always have every hook(even if empty), but it shouldn't be a necessity to have an onExit to move to a state. So yeah, will be changed

Comment (pavelradev812) [reply src/index.js]: I guess that shouldn't be the case, noted.

**Pull request merged**